modules/openapi-generator/src/main/resources/swift5/modelObject.mustache

```diff
@@ -63,7 +63,21 @@
     {{#nonPublicApi}}internal{{/nonPublicApi}}{{^nonPublicApi}}public{{/nonPublicApi}} func encode(to encoder: Encoder) throws {
         var container = encoder.container(keyedBy: CodingKeys.self)
         {{#allVars}}
+        {{#encodeModelNullProperties}}
+        {{#required}}
+        try container.encode{{#isNullable}}IfPresent{{/isNullable}}({{{name}}}, forKey: .{{{name}}})
```

Comment:
@jarrodparkes if something isNullable shouldn't we call `container.encodeNil`?

Reply:
Maybe this should be 
```
        {{#required}}
        {{^isNullable}}
        try container.encode({{{name}}}, forKey: .{{{name}}})
        {{/isNullable}}
        {{#isNullable}}
        if let {{{name}}} = {{{name}}} {
            try container.encode({{{name}}}, forKey: .{{{name}}})
        } else {
            try container.encodeNil(forKey: .{{{name}}})
        }
        {{/isNullable}}
        {{/required}}
```

Reply:
What do you think? This is more a question, than an answer, so I would like to have your feedback on this 🙂

Reply:
@4brunu Your code sample looks right to me, but the problems I seemed to be encountering were weird compound states with `required` and `isNullable`. In fact, these two booleans have confused me (and other devs on my team) quite a bit since they _seem_ to refer to the same thing depending on context.

What I was using as my "source of truth" was regenerate the sample and make sure all optionals are wrapped with the `if/let` construct OR `encodeIfPresent`. If the resulting code ever tried to `encode` an optional, then it wouldn't even compile.

The code I currently seemed to cover these cases well, but completely open for improvement/change. I can try your suggestion and see if the resulting code looks good.

Reply:
`required` means if is must be present, either the value or null.
`isNullable` defines if the type can be null.
https://stackoverflow.com/questions/45575493/what-does-required-in-openapi-really-mean

This makes me rethink the entire PR.

Instead of an addition, this should be a bug fix, and we don't need the new flag `encodeModelNullProperties`.

I think this should be the implementation for all the cases. 
```
        {{#required}}
        {{^isNullable}}
        try container.encode({{{name}}}, forKey: .{{{name}}})
        {{/isNullable}}
        {{#isNullable}}
        if let {{{name}}} = {{{name}}} {
            try container.encode({{{name}}}, forKey: .{{{name}}})
        } else {
            try container.encodeNil(forKey: .{{{name}}})
        }
        {{/isNullable}}
        {{/required}}
        {{^required}}
        try container.encode{{#isNullable}}IfPresent{{/isNullable}}({{{name}}}, forKey: .{{{name}}})
        {{/required}}
```

What do you think?

Reply:
^ I think you're right — I'm going to test this now and get an update on this PR 👍

Reply:
@4brunu Upon closer inspection, I think we can actually simplify further. Here's what I was playing with in a Xcode playground:

```swift
struct Animal: Codable, Hashable {
    var className: String
    var color: String? = "red"
    // ...
}
```

If we exclusively use `encode`, then I believe that is the behavior we want with `required`:

```swift
struct Animal: Codable, Hashable {
    // ...
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(className, forKey: .className)
        try container.encode(color, forKey: .color)
    }
}

func testCodable(obj: Animal) throws {
    let encoder = JSONEncoder()
    let data = try encoder.encode(obj)
    if let jsonString = String(data: data, encoding: .utf8) {
        print(jsonString)
    }
    
    let decoder = JSONDecoder()
    let animal = try decoder.decode(Animal.self, from: data)
    print(animal)
}

do {
    // prints...
    // {"className":"Animal","color":null}
    // Animal(className: "Animal", color: nil)
    try testCodable(obj: Animal(className: "Animal", color: nil))

    // prints...
    // {"className":"Animal","color":"Red"}
    // Animal(className: "Animal", color: Optional("Red"))
    try testCodable(obj: Animal(className: "Animal", color: "Red"))
}
```

If something is not required (`^required`), that is when we should use `encodeIfPresent`:

```swift
struct Animal: Codable, Hashable {
    // ...
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(className, forKey: .className)
        try container.encodeIfPresent(color, forKey: .color)
    }
}

func testCodable(obj: Animal) throws {
    let encoder = JSONEncoder()
    let data = try encoder.encode(obj)
    if let jsonString = String(data: data, encoding: .utf8) {
        print(jsonString)
    }
    
    let decoder = JSONDecoder()
    let animal = try decoder.decode(Animal.self, from: data)
    print(animal)
}

do {
    // prints...
    // {"className":"Animal"}
    // Animal(className: "Animal", color: nil)
    try testCodable(obj: Animal(className: "Animal", color: nil))

    // prints...
    // {"className":"Animal","color":"Red"}
    // Animal(className: "Animal", color: Optional("Red"))
    try testCodable(obj: Animal(className: "Animal", color: "Red"))
}
```

Essentially, `isNullable` shouldn't effect which encode method we use. Only `isRequired`. That would leave us with this...

```
        {{#allVars}}
        try container.encode{{#required}}IfPresent{{/required}}({{{name}}}, forKey: .{{{name}}})
        {{/allVars}}
```

Thoughts?

Reply:
But isn't this the current implementation?

Reply:
![image](https://user-images.githubusercontent.com/1331063/144107601-f0ea3d10-9075-4bee-98ae-7e15887b135b.png)

Reply:
Anyway, thank you for taking the initiate and the time to create a PR to improve the Swift generator, feel free to contribute with PR anytime you want 😃

Reply:
Your conclusion of what works for every situation, is not exactly the same as the current one
Your solution
```
        try container.encode{{#required}}IfPresent{{/required}}({{{name}}}, forKey: .{{{name}}})
```

The current one
```
        try container.encode{{^required}}IfPresent{{/required}}({{{name}}}, forKey: .{{{name}}})
```

But I think the current one is the right one, or am I wrong?

Reply:
@4brunu 

Whoops, I mean it needs to be inverted. The current implementation has it backwards.

```
        {{#allVars}}
        try container.encode{{^required}}IfPresent{{/required}}({{{name}}}, forKey: .{{{name}}})
        {{/allVars}}
```

Reply:
But if it's required, it should always be encoded, even if it's null.
If it's not required, it should be encoded only if it's present, or am I wrong?
So I think the current implementation is correct, or am I mistaken?

Reply:
I think we are both saying the same thing and the implementation is wrong.

> But if it's required, it should always be encoded, even if it's null.
If it's not required, it should be encoded only if it's present, or am I wrong?
So I think the current implementation is correct, or am I mistaken?

_translated_

But if it is `{{#required}}`, then it should always be encoded, even if it's null. ==> `encode`
If it's `{{^required}}`, then it should be encoded only if it's present ==> `encodeIfPresent`

Reply:
We are saying the same thing 🙂 
That's the current implementation on master
```
        try container.encode{{^required}}IfPresent{{/required}}({{{name}}}, forKey: .{{{name}}})
```

Reply:
I think I'm going crazy :face_palm: you're right. I think I was probably just using the combination of `required` and `x-nullable` wrong this whole time

Reply:
It's normal, it's confusing the required vs nullable thing.